scipy/_lib/_array_api.py

```diff
@@ -645,3 +646,16 @@ def xp_float_to_complex(arr: Array, xp: ModuleType | None = None) -> Array:
         arr = xp.astype(arr, xp.complex128)
 
     return arr
+
+def xp_create_diagonal(x: Array, /, *, offset: int = 0,
```

Comment:
I think this should pass, right?
```python3
rng = np.random.default_rng(2347823)
for i in range(1, 10):
    for j in range(1, 10):
        x = rng.random(i)
        A = xp_diag(x, offset=j)
        B = np.diag(x, j)
        assert_allclose(A, B)
```
But there are many assertion and internal value errors. I'd suggest either just including the special case you need in `rosen` or something more obvious. This came to mind and seems to work:

```python3
def xp_diag(x, /, *, offset, xp=np):
    n = x.shape[0] + abs(offset)
    A = np.zeros((n, n))
    i = np.arange(x.shape[0])
    A[abs(offset) + i, i] = x
    return A if offset < 0 else A.T
```

Reply:
Good spot thanks. I couldn't get your implementation to work with array API strict due to the indexing but I fixed my version and added a test

Reply:
I based my version off the numpy source code but agree yours is more compact 
https://github.com/numpy/numpy/blob/v2.1.0/numpy/lib/_twodim_base_impl.py#L247-L317